server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -1492,7 +1511,7 @@ public CommitId flush(boolean force, boolean waitIfOngoing) throws EngineExcepti
                 logger.trace("acquired flush lock immediately");
             }
             try {
-                if (indexWriter.hasUncommittedChanges() || force) {
+                if (indexWriter.hasUncommittedChanges() || force || shouldFlush()) {
```

Comment:
can we add a comment explaining why we have 3 things? Basically something like - we check if:
1) We're forced.
2) There are uncommitted lucene docs in lucene
3) There are translog related reasons to create a new commit which point to a different place in the translog.

Reply:
Done